### PR TITLE
feat(bridge-withdrawer): bridge withdrawer startup

### PR DESCRIPTION
**Pull request by itamarreif**
## Summary
We need to validate config and balance before starting to process transactions. 

1. Added a startup() function to the watcher and submitter each, where set_ready() is called at the end of them. before, we were calling set_ready()  before confirming the config against the node and establishing connection with geth.
2. To keep the sequencer and geth logic separate, there's a one shot channel from the submitter to the watcher that is used for passing the valid fee asset, so the watcher waits for this from the submitter before starting its own startup logic
    a. watcher startup: connecting to geth, confirming contract addr, getting decimals from contract, making the batcher object
    b. submitter startup: confirming chain_id against sequencer node, making sure the fee asset is valid, checking that it meets a minimum balance (also passed from config)
3. refactored some of the submitter unit tests/mock test things to be able to test the startup logic and added a sanity check test + for each of the startup requests failing

## Background
Need to implement startup routine for bridge withdrawer.

## Changes
- [x] submitter
    - [x] confirm chain_id
    - [x] confirm fee asset is valid
    - [x] confirm sufficient balance of fee asset
        - [x] add config var
- [x] watcher
    - [x] get fee asset from submitter?
    - [x] fix connection to geth to use exponential backoff
    - [x] validate decimals against contract

- [x] fix `set_ready()` calls to be at the end of `startup()`

## Testing
- [x] sanity check unit test that the startup works

## Metrics
- List out metrics added by PR, delete section if none. 

## Breaking Changelist
- Bulleted list of breaking changes, any notes on migration. Delete section if none.

## Related Issues
Link any issues that are related, prefer full github links.

closes

### CODE REVIEW

File: crates/astria-bridge-withdrawer/src/config.rs [hunk 1]
```diff
@@ -17,6 +17,9 @@ pub struct Config {
     pub sequencer_key_path: String,
     // The fee asset denomination to use for the bridge account's transactions.
     pub fee_asset_denomination: String,
+    // The minimum expected balance of the fee asset in the bridge account.
+    // TODO: This should be a u128, but the test fails to parse it as such.
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/config.rs:21]: which test?

File: crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs [hunk 2]
```diff
@@ -156,16 +145,90 @@ impl Watcher {
...
+        // get contract handle
+        let contract = AstriaWithdrawer::new(self.contract_address, provider.clone());
+
+        // get asset withdrawal decimals
+        let base_chain_asset_precision = contract
+            .base_chain_asset_precision()
+            .call()
+            .await
+            .wrap_err("failed to get asset withdrawal decimals")?;
+        let asset_withdrawal_divisor = 10u128.pow(base_chain_asset_precision);
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs:217]:
```suggestion
        let asset_withdrawal_divisor = 10u128.pow(18u32.checked_sub(base_chain_asset_precision);
```

File: crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs [hunk 3]
```diff
@@ -156,16 +145,90 @@ impl Watcher {
...
+        AstriaWithdrawerContractHandle,
+        asset::Id,
+        u128,
+    )> {
+        // wait for submitter to be ready
+        let SequencerStartupInfo {
+            fee_asset_id,
+        } = self.submitter_handle.recv_startup_info().await?;
+
+        // connect to geth
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs:177]:
```suggestion
        // connect to eth node
```

File: crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs [hunk 4]
```diff
@@ -156,16 +145,90 @@ impl Watcher {
...
+            .on_retry(
+                |attempt, next_delay: Option<Duration>, error: &ProviderError| {
+                    let wait_duration = next_delay
+                        .map(humantime::format_duration)
+                        .map(tracing::field::display);
+                    warn!(
+                        attempt,
+                        wait_duration,
+                        error = error as &dyn std::error::Error,
+                        "attempt to connect to geth node failed; retrying after backoff",
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs:190]:
```suggestion
                        "attempt to connect to rollup node failed; retrying after backoff",
```

File: crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs [hunk 5]
```diff
@@ -156,16 +145,90 @@ impl Watcher {
...
+        let provider = tryhard::retry_fn(|| {
+            let url = self.ethereum_rpc_endpoint.clone();
+            async move {
+                let websocket_client = Ws::connect_with_reconnects(url, 0).await?;
+                Ok(Provider::new(websocket_client))
+            }
+        })
+        .with_config(retry_config)
+        .await
+        .wrap_err("failed connecting to geth after several retries; giving up")?;
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs:205]:
```suggestion
        .wrap_err("failed connecting to rollup node after several retries; giving up")?;
```

File: crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs [hunk 6]
```diff
@@ -450,19 +520,26 @@ mod tests {
         };
         let denom: Denom = Denom::from_base_denom("nria");
         let expected_action =
-            event_to_action(expected_event, denom.id(), denom.clone(), 1).unwrap();
+            event_to_action(expected_event, denom.id(), denom.clone(), 10u128.pow(18)).unwrap();
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs:523]:
```suggestion
            event_to_action(expected_event, denom.id(), denom.clone(), 1).unwrap();
```

after the logic for the divisor above is fixed

File: crates/astria-bridge-withdrawer/src/withdrawer/submitter/builder.rs [hunk 7]
```diff
@@ -1,46 +1,92 @@
...
-use crate::withdrawer::submitter::Batch;
+use crate::withdrawer::{
+    submitter::Batch,
+    SequencerStartupInfo,
+};
 
 const BATCH_QUEUE_SIZE: usize = 256;
 
+pub(crate) struct Handle {
+    startup_info_rx: Option<oneshot::Receiver<SequencerStartupInfo>>,
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/withdrawer/submitter/builder.rs:24]: don't think we need an option here - if the oneshot returns an error, then `recv_startup_info` should also error

File: crates/astria-bridge-withdrawer/src/withdrawer/submitter/mod.rs [hunk 8]
```diff
@@ -58,17 +74,15 @@ pub(super) struct Submitter {
     sequencer_cometbft_client: sequencer_client::HttpClient,
     signer: SequencerKey,
     sequencer_chain_id: String,
+    startup_tx: Option<oneshot::Sender<SequencerStartupInfo>>,
```
Comment (noot) [inline crates/astria-bridge-withdrawer/src/withdrawer/submitter/mod.rs:77]: same here, don't think we need option

**Review by noot — CHANGES_REQUESTED** [review]
please fix `asset_withdrawal_divisor` calculation, otherwise nothing major!

**Review by noot — APPROVED** [review]
lgtm!

**Comment by SuperFluffy** [comment]
All these startup logics completely ignore shutdown signals.

**Review by joroshiba — APPROVED** [review]
infra approval, lean on janis & elizabeth for rust side.

**Pull request merged**